crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -0,0 +1,395 @@
+            // avoid searching the cached timeline events.
+            trace!("We got a new event with a read receipt: {receipt_event_id}. Search in new events...");
+            if find_and_count_events(&receipt_event_id, user_id, new_events.iter(), room_info) {
+                // It did, so our work here is done.
+                return Ok(());
+            }
+
+            // We didn't find the event attached to the receipt in the new batches of
+            // events. It's possible it's referring to an event we've already
+            // seen. In that case, try to find it.
+            let previous_events = previous_events_provider.for_room(&room_info.room_id);
+
+            trace!("Couldn't find the event attached to the receipt in the new events; looking in past events too now...");
+            if find_and_count_events(
+                &receipt_event_id,
+                user_id,
+                previous_events.iter().chain(new_events.iter()),
+                room_info,
+            ) {
+                // It did refer to an old event, so our work here is done.
+                return Ok(());
+            }
+        }
+    }
+
+    if let Some(receipt_event_id) = prev_latest_receipt_event_id {
+        // There's no new read-receipt here. We assume the cached events have been
+        // properly processed, and we only need to process the new events based
+        // on the previous receipt.
+        trace!("Couldn't find the event attached to the latest receipt; looking if the past latest known receipt refers to a new event...");
+        if find_and_count_events(&receipt_event_id, user_id, new_events.iter(), room_info) {
+            // We found the event to which the previous receipt attached to, so our work is
+            // done here.
+            return Ok(());
+        }
+    }
+
+    // If we haven't returned at this point, it means that either we had no previous
+    // read receipt, or the previous read receipt was not attached to any new
+    // event.
+    //
+    // In that case, accumulate all events as part of the current batch, and wait
+    // for the next receipt.
+    trace!("All other ways failed, including all new events for the receipts count.");
+    for event in new_events {
+        if event.push_actions.iter().any(ruma::push::Action::is_highlight) {
+            room_info.notification_counts.highlight_count += 1;
+        }
+        if marks_as_unread(&event.event, user_id) {
+            room_info.notification_counts.notification_count += 1;
+        }
+    }
+
+    Ok(())
+}
+
+/// Try to find the event to which the receipt attaches to, and if found, will
+/// update the notification count in the room.
+///
+/// Returns a boolean indicating if it's found the event and updated the count.
+fn find_and_count_events<'a>(
+    receipt_event_id: &EventId,
+    user_id: &UserId,
+    events: impl Iterator<Item = &'a SyncTimelineEvent>,
```

Comment:
You maybe want to use `IntoIterator<Item = &'a SyncTimelineEvent>`, so that it simplifies some calls to `find_and_count_events`. For example, it might replace:

```rust
find_and_count_events(&receipt_event_id, user_id, new_events.iter(), room_info)
```

to:

```rust
find_and_count_events(&receipt_event_id, user_id, new_events, room_info)
```

Thoughts?